x-pack/plugins/cases/public/components/configure_cases/index.tsx

```diff
@@ -312,24 +357,141 @@ export const ConfigureCases: React.FC = React.memo(() => {
     ]
   );
 
-  const CustomFieldAddFlyout = customFieldFlyoutVisible ? (
-    <CustomFieldFlyout
-      isLoading={loadingCaseConfigure || isPersistingConfiguration}
-      disabled={
-        !permissions.create ||
-        !permissions.update ||
-        loadingCaseConfigure ||
-        isPersistingConfiguration
-      }
-      customField={customFieldToEdit}
-      onCloseFlyout={onCloseAddFieldFlyout}
-      onSaveField={onSaveCustomField}
-    />
-  ) : null;
+  const onCloseTemplateFlyout = useCallback(() => {
+    setFlyOutVisibility({ type: 'template', visible: false });
+    setTemplateToEdit(null);
+  }, [setFlyOutVisibility, setTemplateToEdit]);
+
+  const onTemplateSave = useCallback(
+    (data: TemplateFormProps | null) => {
+      const {
+        connectorId,
+        fields,
+        customFields: templateCustomFields,
+        syncAlerts = false,
+        key,
+        name,
+        templateTags,
+        templateDescription,
+        ...otherCaseFields
+      } = (data ?? {}) as TemplateFormProps;
+
+      const transformedCustomFields = templateCustomFields
+        ? transformCustomFieldsData(templateCustomFields, customFields)
+        : [];
+      const templateConnector = connectorId ? getConnectorById(connectorId, connectors) : null;
+
+      const transformedConnector = templateConnector
+        ? normalizeActionConnector(templateConnector, fields)
+        : getNoneConnector();
+
+      const transformedData: TemplateConfiguration = {
+        key,
+        name,
+        description: templateDescription,
+        tags: templateTags ?? [],
+        caseFields: {
+          ...otherCaseFields,
+          connector: transformedConnector,
+          customFields: transformedCustomFields,
+          settings: { syncAlerts },
+        },
+      };
+
+      const updatedTemplates = addOrReplaceField(templates, transformedData);
+
+      persistCaseConfigure({
+        connector,
+        customFields,
+        templates: updatedTemplates,
+        id: configurationId,
+        version: configurationVersion,
+        closureType,
+      });
+
+      setFlyOutVisibility({ type: 'template', visible: false });
+      setTemplateToEdit(null);
+    },
+    [
+      closureType,
+      configurationId,
+      configurationVersion,
+      connector,
+      connectors,
+      customFields,
+      templates,
+      persistCaseConfigure,
+    ]
+  );
+
+  const AddOrEditCustomFieldFlyout = useMemo(() => {
+    return flyOutVisibility?.type === 'customField' && flyOutVisibility?.visible ? (
+      <CommonFlyout<CustomFieldConfiguration>
+        isLoading={loadingCaseConfigure || isPersistingConfiguration}
+        disabled={
+          !permissions.create ||
+          !permissions.update ||
+          loadingCaseConfigure ||
+          isPersistingConfiguration
+        }
+        onCloseFlyout={onCloseCustomFieldFlyout}
+        onSaveField={onCustomFieldSave}
+        data={customFieldToEdit as CustomFieldConfiguration}
```

Comment:
removed all type casting from [this file](https://github.com/elastic/kibana/pull/184104/commits/83796795d869900f20aad879f0d020a4a0d773a0#diff-a6135a7ecd6b7f879370d40b47f78a14c6715335bfa8226a77cd433d40311564)

Except templateForm's initialValue.  Because here templateToEdit is of type `TemplateConfiguration | null` and will be converted to `TemplateFormProps` using a deSerializer in the next PR.

```
 <TemplateForm initialValue={templateToEdit as TemplateFormProps | null} ... />
```